dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,670 @@
+	// requests from that class.
 	//
-	// +required
 	// +listType=atomic
-	Devices []DeviceClaimInstance `json:"devices,omitempty"`
+	Requests []ResourceRequest `json:"requests,omitempty" protobuf:"bytes,5,name=requests"`
+
+	// Future extension, ignored by older schedulers. This is fine because scoring
+	// allows users to define a preference, without making it a hard requirement.
+	//
+	//
+	// Score *SomeScoringStruct
+
+	// Shareable indicates whether the allocated claim is meant to be shareable
+	// by multiple consumers at the same time.
+	// +optional
+	Shareable bool `json:"shareable,omitempty" protobuf:"bytes,3,opt,name=shareable"`
 }
 
-// DeviceClaimInstance captures a group of claims which must all be satisfied,
-// and a group for which one must be sastisfied. If you require multiple
-// groups, one of each may be satsfied, you will need multiple instances.
-type DeviceClaimInstance struct {
-	// Name is used to associate this instance with a specific
-	// container.
-	Name string `json:"name"`
+// ClassReference must have one and only one field set.
+//
+// If we ever need to introduce new class settings which cannot be added to ResourceClass
+// (for example, because it wouldn't be okay for an older scheduler to ignore them),
+// then we could add new types or entries here to indicate to old schedulers that
+// they cannot handle the claim. At that point we can even rename the "Name" field,
+// if a longer name then makes more sense (conversion could handle the difference).
+type ClassReference struct {
+	// When referencing a ResourceClass, a claim inherits additional
+	// configuration, constraints and requirements.
+	//
+	// If the claim contains no requests, the first non-empty default
+	// requests defined by one of these classes are used. If no class
+	// provides such defaults, then a single, empty request is used.
+	//
+	// +optional
+	// +listType=atomic
+	ResourceClassName *string `json:"resourceClassName,omitempty"`
+}
 
-	// Previously, a DeviceClaimInstance resolved to a single
-	// DeviceClaimDetail, so we didn't need matchAttributes here. With
-	// this version, we *could* use matchAttributes here, but most use
-	// cases are likely satisfied by the matchAttributes at the Spec level,
-	// so for now we leave it out of this level.
+// ResourceRequest is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for one of several different alternatives.
+type ResourceRequest struct {
+	// Name can be set to enable referencing this request in a pod.spec.containers[].resources.devices
+	// entry, if that is desired.
+	//
+	// +optional
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
 
-	// AllOf contains a list of claims, all of which must be satisfied.
-	// be empty.
-	AllOf []DeviceClaimDetail `json:"allOf,omitempty"`
+	*ResourceRequestDetail `json:",inline,omitempty"`
```

Comment:
I had this initially too but Tim considered it suspect from an API review standpoint, so I shifted to the `allOf`.

Reply:
We have a list of requests, therefore I didn't see a need to have another repetition here with an `allOf``.

`oneOf` is needed to express the intended preference ("first this, otherwise this, etc.").

Reply:
The inlined pointer struct is novel, so it wasn't clear api reviewers would be ok with it.

Reply:
I think I saw Tim's question about this on Slack and the response was "it's okay".